{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c219c1a2-cdca-48ad-bb00-b692a14b2f4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "tqdm.pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e4ecc29b-2077-410d-8137-ee48d4a294d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = \"gencode.v22.annotation.gtf.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6b67ee5e-d8cd-46f7-868c-8803186bb14f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>annotation_source</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>genomic_start_location</th>\n",
       "      <th>genomic_end_location</th>\n",
       "      <th>additional_info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; gene_type \"transc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563666</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>gene_id \"ENSG00000210195.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563667</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>gene_id \"ENSG00000210195.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563668</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; gene_type \"Mt_tRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563669</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563670</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2563671 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chromosome_name annotation_source feature_type  \\\n",
       "0                  chr1            HAVANA         gene   \n",
       "1                  chr1            HAVANA   transcript   \n",
       "2                  chr1            HAVANA         exon   \n",
       "3                  chr1            HAVANA         exon   \n",
       "4                  chr1            HAVANA         exon   \n",
       "...                 ...               ...          ...   \n",
       "2563666            chrM           ENSEMBL   transcript   \n",
       "2563667            chrM           ENSEMBL         exon   \n",
       "2563668            chrM           ENSEMBL         gene   \n",
       "2563669            chrM           ENSEMBL   transcript   \n",
       "2563670            chrM           ENSEMBL         exon   \n",
       "\n",
       "         genomic_start_location  genomic_end_location  \\\n",
       "0                         11869                 14409   \n",
       "1                         11869                 14409   \n",
       "2                         11869                 12227   \n",
       "3                         12613                 12721   \n",
       "4                         13221                 14409   \n",
       "...                         ...                   ...   \n",
       "2563666                   15888                 15953   \n",
       "2563667                   15888                 15953   \n",
       "2563668                   15956                 16023   \n",
       "2563669                   15956                 16023   \n",
       "2563670                   15956                 16023   \n",
       "\n",
       "                                           additional_info  \n",
       "0        gene_id \"ENSG00000223972.5\"; gene_type \"transc...  \n",
       "1        gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  \n",
       "2        gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  \n",
       "3        gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  \n",
       "4        gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  \n",
       "...                                                    ...  \n",
       "2563666  gene_id \"ENSG00000210195.2\"; transcript_id \"EN...  \n",
       "2563667  gene_id \"ENSG00000210195.2\"; transcript_id \"EN...  \n",
       "2563668  gene_id \"ENSG00000210196.2\"; gene_type \"Mt_tRN...  \n",
       "2563669  gene_id \"ENSG00000210196.2\"; transcript_id \"EN...  \n",
       "2563670  gene_id \"ENSG00000210196.2\"; transcript_id \"EN...  \n",
       "\n",
       "[2563671 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\n",
    "        file, sep=\"\\t\", comment=\"#\", compression='gzip', \n",
    "        usecols=[0, 1, 2, 3, 4, 8],\n",
    "        header=None, names=(\n",
    "            \"chromosome_name\", \"annotation_source\", \"feature_type\",\n",
    "            \"genomic_start_location\", \"genomic_end_location\", \n",
    "            # \"score\", \"genomic_strand\", \"genomic_phase\", \n",
    "            \"additional_info\"\n",
    "            )\n",
    "    )\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a234d84b-6750-49a0-af68-edc1620b7cd5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████████████████████████████| 2563671/2563671 [02:13<00:00, 19139.68it/s]\n"
     ]
    }
   ],
   "source": [
    "def parseinfo(s):\n",
    "    \n",
    "    items = s['additional_info'].replace(\" \", \"\").split(\";\")\n",
    "    gene_id = None\n",
    "    gene_type = None\n",
    "    gene_name = None\n",
    "    for i in items:\n",
    "        if 'gene_id' in i:\n",
    "            gene_id = i.split('\"')[1]\n",
    "        elif 'gene_type' in i:\n",
    "            gene_type = i.split('\"')[1]\n",
    "        elif 'gene_name' in i:\n",
    "            gene_name = i.split('\"')[1]\n",
    "    return gene_id, gene_type, gene_name\n",
    "\n",
    "df[['ensembl_id', 'gene_type', 'gene_name']] = df[['additional_info']].progress_apply(parseinfo, axis=1, result_type='expand')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "de1b8fb2-54de-479e-8dba-03100b93ee8c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>annotation_source</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>genomic_start_location</th>\n",
       "      <th>genomic_end_location</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563666</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563667</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563668</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563669</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563670</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2563671 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chromosome_name annotation_source feature_type  \\\n",
       "0                  chr1            HAVANA         gene   \n",
       "1                  chr1            HAVANA   transcript   \n",
       "2                  chr1            HAVANA         exon   \n",
       "3                  chr1            HAVANA         exon   \n",
       "4                  chr1            HAVANA         exon   \n",
       "...                 ...               ...          ...   \n",
       "2563666            chrM           ENSEMBL   transcript   \n",
       "2563667            chrM           ENSEMBL         exon   \n",
       "2563668            chrM           ENSEMBL         gene   \n",
       "2563669            chrM           ENSEMBL   transcript   \n",
       "2563670            chrM           ENSEMBL         exon   \n",
       "\n",
       "         genomic_start_location  genomic_end_location         ensembl_id  \\\n",
       "0                         11869                 14409  ENSG00000223972.5   \n",
       "1                         11869                 14409  ENSG00000223972.5   \n",
       "2                         11869                 12227  ENSG00000223972.5   \n",
       "3                         12613                 12721  ENSG00000223972.5   \n",
       "4                         13221                 14409  ENSG00000223972.5   \n",
       "...                         ...                   ...                ...   \n",
       "2563666                   15888                 15953  ENSG00000210195.2   \n",
       "2563667                   15888                 15953  ENSG00000210195.2   \n",
       "2563668                   15956                 16023  ENSG00000210196.2   \n",
       "2563669                   15956                 16023  ENSG00000210196.2   \n",
       "2563670                   15956                 16023  ENSG00000210196.2   \n",
       "\n",
       "                                  gene_type gene_name  \n",
       "0        transcribed_unprocessed_pseudogene   DDX11L1  \n",
       "1        transcribed_unprocessed_pseudogene   DDX11L1  \n",
       "2        transcribed_unprocessed_pseudogene   DDX11L1  \n",
       "3        transcribed_unprocessed_pseudogene   DDX11L1  \n",
       "4        transcribed_unprocessed_pseudogene   DDX11L1  \n",
       "...                                     ...       ...  \n",
       "2563666                             Mt_tRNA     MT-TT  \n",
       "2563667                             Mt_tRNA     MT-TT  \n",
       "2563668                             Mt_tRNA     MT-TP  \n",
       "2563669                             Mt_tRNA     MT-TP  \n",
       "2563670                             Mt_tRNA     MT-TP  \n",
       "\n",
       "[2563671 rows x 8 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df = df[[\"chromosome_name\", \"annotation_source\", \"feature_type\",\n",
    "            \"genomic_start_location\", \"genomic_end_location\", \n",
    "            # \"score\", \"genomic_strand\", \"genomic_phase\", \n",
    "            \"ensembl_id\", \"gene_type\", \"gene_name\"]]\n",
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "09f47dac-fbd8-40cd-9636-805b5e5b1390",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████████████████████████████████████████| 60483/60483 [00:43<00:00, 1403.04it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.13</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>4535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <td>TNMD</td>\n",
       "      <td>1610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.11</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>1207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.12</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>6883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.15</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>5967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSGR0000275287.3</th>\n",
       "      <td>Metazoa_SRP</td>\n",
       "      <td>290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSGR0000276543.3</th>\n",
       "      <td>AJ271736.1</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSGR0000277120.3</th>\n",
       "      <td>MIR6089</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSGR0000280767.1</th>\n",
       "      <td>RP13-465B17.5</td>\n",
       "      <td>515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSGR0000281849.1</th>\n",
       "      <td>RP13-465B17.4</td>\n",
       "      <td>484</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60483 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           symbol  length\n",
       "ensembl_id                               \n",
       "ENSG00000000003.13         TSPAN6    4535\n",
       "ENSG00000000005.5            TNMD    1610\n",
       "ENSG00000000419.11           DPM1    1207\n",
       "ENSG00000000457.12          SCYL3    6883\n",
       "ENSG00000000460.15       C1orf112    5967\n",
       "...                           ...     ...\n",
       "ENSGR0000275287.3     Metazoa_SRP     290\n",
       "ENSGR0000276543.3      AJ271736.1      68\n",
       "ENSGR0000277120.3         MIR6089      64\n",
       "ENSGR0000280767.1   RP13-465B17.5     515\n",
       "ENSGR0000281849.1   RP13-465B17.4     484\n",
       "\n",
       "[60483 rows x 2 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def unioninterval(l):\n",
    "    \n",
    "    def get_union(a, b):\n",
    "        a0, a1 = a\n",
    "        b0, b1 = b\n",
    "        if a1 + 1 == b0:\n",
    "            return [1, a0, b1]\n",
    "        if a1 < b0:\n",
    "            return [2, a, b]\n",
    "        if a0 <= b0:\n",
    "            return [1, a0, max(a1, b1)]\n",
    "    \n",
    "    l.sort(key=lambda x: x[0])\n",
    "    i, j = 0, len(l) - 1\n",
    "    while True:\n",
    "        if i == j:\n",
    "            break\n",
    "        rangea, rangeb = l.pop(i), l.pop(i)\n",
    "        union_i = get_union(rangea, rangeb)\n",
    "        if union_i[0] == 1:\n",
    "            l.insert(i, [union_i[1], union_i[2]])\n",
    "        elif union_i[0] == 2:\n",
    "            l.insert(i, union_i[1])\n",
    "            l.insert(i + 1, union_i[2])\n",
    "            i += 1\n",
    "        j = len(l) - 1\n",
    "    return l\n",
    "\n",
    "def addup(sdf):\n",
    "    interval = unioninterval(\n",
    "        sdf[['genomic_start_location', 'genomic_end_location']].values.tolist()\n",
    "    )\n",
    "    length = 0\n",
    "    for x in interval:\n",
    "        length += x[1] - x[0] + 1        \n",
    "    return pd.Series([sdf.iloc[0, :]['gene_name'], length])\n",
    "\n",
    "exon_df = new_df.loc[new_df['feature_type'] == 'exon', ].groupby('ensembl_id').progress_apply(addup)\n",
    "exon_df.columns = ['symbol', 'length']\n",
    "exon_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c868dd6b-3677-4d6d-9759-d490d7935f6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_df.to_csv('gene.v22.simple.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
